fix: complete toc disambiguation

dandhlee opened the pull request:
Before you open a pull request, note that this repository is forked from [here](https://github.com/docascode/sphinx-docfx-yaml/).
Unless the issue you're trying to solve is unique to this specific repository, 
please file an issue and/or send changes upstream to the original as well.

__________________________________________________________________

In #16 , it looked for cases which has `google.cloud.package.module` to shorten the entries, and if there are multiple `modules` then we relied on using different package names. However this did not work if there were more parts involved, and if the unique disambiguation we were looking for fell in the middle.

To fix this, for all entries at the same level I've created a map of all words separated by `.` and determine whether there are unique entries of the word or not. It won't matter if there are only single entries of each entries for the left nav. If there are multiple entries, we look for unique entry + name to use in the left nav. Since there is only one unique identifier for each entry, that should be used along with the name to completely disambiguate given a list.

This should now fix all issues with duplicated entries in the left nav!

- [x] Tests pass
- [ ] Appropriate changes to README are included in PR

dandhlee commented:
:O neat, the Kokoro job gets linked here if I run based on this PR's git hash :D (it failed because I deleted a blob mid-run for a job, ran another job and it should pass.)

tbpg commented on docfx_yaml/extension.py:
```diff
@@ -720,23 +720,42 @@ def sanitize_uidname_field(toc_yaml):
-    # Parses the package name and returns package name and module name.
-    def find_package_name(package_name):
+    # Parses the package name and returns unique identifer and name.
+    def find_unique_name(package_name, entries):
```
I hate to say, this is begging for a unit test. It appears to be self contained and we know exactly what the output should be based on the given input?

dandhlee replied:
Could this wait until #44 ? I'm planning on addressing all of this with that.

tbpg replied:
I don't think so. This wouldn't be a test for verifying output, it would be a test of this function.

Right now, reviewing this, it seems OK. But, a test would make me feel much better about it.

dandhlee replied:
Sounds good.

tbpg commented on docfx_yaml/extension.py:
```diff
@@ -720,23 +720,42 @@ def sanitize_uidname_field(toc_yaml):
+            # Split the name and mark all duplicates.
+            # There will be at least 1 unique identifer for each name.
+            for part in module['uidname'].split("."):
+                names[module_name][part] = 1 if part not in names[module_name] else 2
```
Should this increment instead of just using 2? I was confused at first.

dandhlee replied:
Updated. Since the count just has to be greater than 1 and that specific `x = val_1 if condition else val_2` syntax doesn't work with incrementing counter I've just wanted to keep it at 2, but it would be better to increase readability over few lines of code.

tbpg commented on docfx_yaml/extension.py:
```diff
@@ -720,23 +720,45 @@ def sanitize_uidname_field(toc_yaml):
 
     # Used to disambiguate names that have same entries.
     def disambiguate_toc_name(toc_yaml):
         names = {}
```
Consider a different name from `names`, as this holds the number of times each name occurs, not just the names. `name_counts`?

dandhlee replied:
changed it to `name_entries` since we call its list `entries` above in the `find_package_name` function!

dandhlee commented:
* Updated to move the functions out to easily have them tested with unittest
* Added `test_unit.py` file for unit tests, as well as some yaml test files
* Added `tox -e unittest` for the GitHub Action to use and test (already reflected in the checks)

Ready for review! Please take a look again :)

tbpg commented on tests/test_unit.py:
```diff
@@ -0,0 +1,48 @@
+        entries = {}
+        
+        # Disambiguate with unique entries.
+        entry1_1 = "google.cloud.aiplatform.v1.schema.predict.instance_v1.types"
+        entry1_2 = "google.cloud.aiplatform.v1beta2.schema.predict.instance_v1.types"
+        want1_1 = "v1.types"
+        want1_2 = "v1beta2.types"
```
Why the `1`? (e.g. entry**1**_1).

dandhlee replied:
I had few more sets of tests, then realized one will be good enough to test but forgot to remove the grouping prefix. Removed it.

tbpg commented on tests/test_unit.py:
```diff
@@ -0,0 +1,48 @@
+
+    def test_disambiguate_toc_name(self):
+
+        want_file = open('tests/yaml_post.yaml', 'r')
```
Can you shorten these test files to be 10s of lines that are easy to understand, rather than ~1100 that are difficult to compare?

dandhlee replied:
done!

tbpg reviewed (changes requested):
Thank you!

tbpg reviewed (approved):
Thank you! And thank you for adding tests. :smile:
